crates/oxc_linter/src/disable_directives.rs

```diff
@@ -0,0 +1,257 @@
+use oxc_ast::{Span, Trivias};
+use rust_lapper::{Interval, Lapper};
+use rustc_hash::FxHashMap;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+enum DisabledRule<'a> {
+    All,
+    Single(&'a str),
+}
+
+#[derive(Debug)]
+pub struct DisableDirectives<'a> {
+    /// All the disabled rules with their corresponding covering spans
+    intervals: Lapper<u32, DisabledRule<'a>>,
+}
+
+impl<'a> DisableDirectives<'a> {
+    pub fn contains(&self, rule_name: &'static str, start: u32) -> bool {
+        self.intervals.find(start, start + 1).any(|interval| {
+            interval.val == DisabledRule::All || interval.val == DisabledRule::Single(rule_name)
+        })
+    }
+}
+
+pub struct DisableDirectivesBuilder<'a, 'b> {
+    source_text: &'a str,
+    trivias: &'b Trivias,
+    /// All the disabled rules with their corresponding covering spans
+    intervals: Lapper<u32, DisabledRule<'a>>,
+    /// Start of `eslint-disable`
+    disable_all_start: Option<u32>,
+    /// Start of `eslint-disable rule_name`
+    disable_start_map: FxHashMap<&'a str, u32>,
+}
+
+impl<'a, 'b> DisableDirectivesBuilder<'a, 'b> {
+    pub fn new(source_text: &'a str, trivias: &'b Trivias) -> Self {
+        Self {
+            source_text,
+            trivias,
+            intervals: Lapper::new(vec![]),
+            disable_all_start: None,
+            disable_start_map: FxHashMap::default(),
+        }
+    }
+
+    pub fn build(mut self) -> DisableDirectives<'a> {
+        self.build_impl();
+        DisableDirectives { intervals: self.intervals }
+    }
+
+    fn add_interval(&mut self, start: u32, stop: u32, val: DisabledRule<'a>) {
+        self.intervals.insert(Interval { start, stop, val });
+    }
+
+    #[allow(clippy::cast_possible_truncation)] // for `as u32`
+    fn build_impl(&mut self) {
+        let source_len = self.source_text.len() as u32;
+        // This algorithm iterates through the comments and builds all intervals
+        // for matching disable and enable pairs.
+        // Wrongly ordered matching pairs are not taken into considering.
+        for (start, comment) in self.trivias.comments() {
+            let span = Span::new(*start, comment.end());
+            let text = span.source_text(self.source_text);
+            let text = text.trim_start();
+
+            if let Some(text) = text.strip_prefix("eslint-disable") {
+                // `eslint-disable`
+                if text.trim().is_empty() {
+                    self.disable_all_start = Some(span.end);
+                    continue;
+                }
+
+                // `eslint-disable-next-line`
+                if let Some(_text) = text.strip_prefix("-next-line") {
+                    // Get the span between the next two newlines for "next line"
+                    let mut newlines = self.trivias.newlines().range(span.end..);
```

Comment:
I have a feeling that scanning for the next two newlines in place will be faster overall, given the fact that saving all newlines has a larger overhead.